File: services/server/src/protocol/protocol_test.py
import unittest

from protocol import Protocol
from lottery.bet import Bet


class TestProtocol(unittest.TestCase):
    def test_serialize_deserialize_bets(self):
        sent = [
            Bet(1, "Martino", "Nervi", 12345678, "1999-03-01", 7574),
            Bet(1, "Cirilo", "Pato", 91011113, "2004-05-10", 1033),
        ]
        protocol = Protocol(None)

        payload = len(sent).to_bytes(2, "big")
        for bet in sent:
            payload += protocol._serialize_bet(bet)

        received = protocol._deserialize_bets(payload, 1)

        self.assertEqual(received, sent)


if __name__ == "__main__":
    unittest.main()

File: services/client/src/protocol/protocol_test.go
package protocol

import (
	"testing"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
)

func TestSerializeDeserializeBets(t *testing.T) {
	sent := []lottery.Bet{
		{AgencyId: 1, FirstName: "Martino", LastName: "Nervi", Document: 12345678, Birthdate: "1999-03-01", Number: 7574},
		{AgencyId: 1, FirstName: "Cirilo", LastName: "Pato", Document: 91011113, Birthdate: "2004-05-10", Number: 9325},
	}

	payload := []byte{byte(len(sent))}
	for _, bet := range sent {
		buffer, err := serializeBet(bet)
		if err != nil {
			t.Fatalf("serializeBet error: %v", err)
		}
		payload = append(payload, buffer...)
	}

	received, err := deserializeBets(payload)
	if err != nil {
		t.Fatalf("deserializeBets error: %v", err)
	}
	if len(received) != len(sent) {
		t.Fatalf("esperaba %d apuestas, obtuve %d", len(sent), len(received))
	}
	for i := range sent {
		if received[i] != sent[i] {
			t.Errorf("bet[%d]:\n  got  %+v\n  want %+v", i, received[i], sent[i])
		}
	}
}
